flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java

```diff
@@ -508,18 +517,40 @@ private static void start(YarnConfiguration conf, String principal, String keyta
 		}
 		System.setProperty("user.home", homeDir.getAbsolutePath());
 		String uberjarStartLoc = "..";
+
+		// find flink-dist jar in flink-dist module
 		LOG.info("Trying to locate uberjar in {}", new File(uberjarStartLoc));
-		flinkUberjar = findFile(uberjarStartLoc, new RootDirFilenameFilter());
-		Assert.assertNotNull("Flink uberjar not found", flinkUberjar);
-		String flinkDistRootDir = flinkUberjar.getParentFile().getParent();
-		flinkLibFolder = flinkUberjar.getParentFile(); // the uberjar is located in lib/
-		Assert.assertNotNull("Flink flinkLibFolder not found", flinkLibFolder);
+		final File originalFlinkDistJar = findFile(uberjarStartLoc, new RootDirFilenameFilter());
+
+		// copy entirety of distribution into a temporary location
+		final Path originalFlinkDistRootDir = originalFlinkDistJar.getParentFile().getParentFile().toPath();
+		System.out.println("dist=" + originalFlinkDistRootDir);
+		final Path flinkDistRootDir = tmp.newFolder("tmp_dist_directory").toPath();
+
+		FileUtils.copyDirectory(originalFlinkDistRootDir.toFile(), flinkDistRootDir.toFile());
+
+		flinkLibFolder = flinkDistRootDir.resolve("lib").toFile();
 		Assert.assertTrue("lib folder not found", flinkLibFolder.exists());
 		Assert.assertTrue("lib folder not found", flinkLibFolder.isDirectory());
+		try (Stream<Path> libJars = Files.list(flinkLibFolder.toPath())) {
+			final Optional<File> flinkDistJarOptional =
+				libJars.map(Path::getFileName)
+					.map(Path::toString)
+					.filter(RootDirFilenameFilter::isFlinkDistJar)
+					.map(fileName -> flinkLibFolder.toPath().resolve(Paths.get(fileName)))
+					.map(Path::toFile)
+					.findAny();
+			flinkUberjar = flinkDistJarOptional.orElseThrow(() -> new AssertionError("Unable to locate flink-dist jar."));
+		}
 
-		if (!flinkUberjar.exists()) {
-			Assert.fail("Unable to locate yarn-uberjar.jar");
+		// copy flink-shaded-hadoop2 into dist, since it is not included by default
+		// the hadoop jar was copied into the dependencies directory during the build using the maven-dependency-plugin
+		final Path relHadoopPath;
+		try (Stream<Path> dependencyJars = Files.list(Paths.get("target/dependencies"))) {
+			relHadoopPath = dependencyJars.filter(jar -> jar.getFileName().toString().startsWith("flink-shaded-hadoop2"))
+				.findAny().orElseThrow(() -> new AssertionError("Unable to locate flink-shaded-hadoop2 jar."));
 		}
+		Files.copy(relHadoopPath, flinkLibFolder.toPath().resolve("flink-shaded-hadoop2.jar"));
```

Comment:
Do we strictly need to copy everything together? Wouldn't it also work to refactor the `YarnTestBase` a bit so that we have a method `getShipFiles` which returns all files in `/lib` plus the `flink-shaded-hadoop2.jar`? That way we would save a lot of copy operations. `flink-dist` is currently 347 MB large.

Reply:
That could probably work, but you'd need to give me more hints how this could be implemented.

Reply:
I think it is mainly adding the `flink-shaded-hadoop2.jar` as a shippable file via `-yt` in the Yarn tests. With https://issues.apache.org/jira/browse/FLINK-11272 this should be possible by adding `-yt path_to_shaded_hadoop` to the test cases which use the Yarn cli. If you want, I can also take over to do it.

Reply:
Doesn't the client require hadoop as well? `-yt` doesn't automatically put the passed jar on the classpath of the client, does it?